src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,368 @@
+            raise TypeError(f"Expected config type to be 'XLoraConfig', got '{type(model)}' instead.")
+
+        self.xlora_config = peft_config
+
+        if hasattr(model.config, "use_cache"):
+            assert not model.config.use_cache, "`use_cache` must be False"
+
+        use_trainable_adapters = peft_config.use_trainable_adapters
+        adapters_items = iter(peft_config.adapters.items())
+
+        # For load_adapter to think we are a LoraModel
+        model_peft.peft_type = PeftType.LORA
+
+        for name, model_id in adapters_items:
+            model_peft.load_adapter(model_id, name, is_trainable=use_trainable_adapters)
+
+        self.delete_adapter(adapter_name)
+
+        self.set_adapter(list(peft_config.adapters.keys()))
+        model_peft.active_adapter = name
+        model_peft.peft_type = PeftType.XLORA
+
+        def hook(module, *args, **kwargs) -> None:
+            args_real = args[0]
+            kwargs_real: dict = args[1]
+            kwargs_real.update(kwargs)
+
+            xlora_classifier: XLoraClassifier = self.internal_xlora_classifier  # type: ignore
+
+            if "_xlora_classifier_inhibitor_flag" in kwargs_real:
+                payload: InhibitorFlagPayload = kwargs_real["_xlora_classifier_inhibitor_flag"]
+
+                del kwargs_real["_xlora_classifier_inhibitor_flag"]
+
+                self.internal_xlora_scalings = torch.full(  # type: ignore
+                    (payload.batch_size, payload.seq_len, xlora_classifier.n_layers, xlora_classifier.n_classes),
+                    payload.override_scaling_pass_value,
+                )
+
+                return
+
+            xlora_scalings = xlora_classifier.forward(
+                *args_real,
+                **kwargs_real,
+            )
+            # Set the scalings
+            self.internal_xlora_scalings = xlora_scalings
+
+        model.register_forward_pre_hook(hook, with_kwargs=True, prepend=True)
+
+        self._freeze_all_adapters()
+
+        total_swapped = convert_layers_to_xlora(
+            model_peft,
+            self,
+            peft_config,
+        )
+
+        n_classes = len(peft_config.adapters)
+        xlora_classifier = XLoraClassifier(model_peft, peft_config, n_classes, total_swapped)
+
+        # Setup the model internal state
+        self.__dict__["internal_xlora_classifier"] = xlora_classifier
+        self.internal_xlora_scalings = None  # type: ignore
+
+    def _freeze_all_adapters(self):
+        self.eval()
+        if not self.xlora_config.use_trainable_adapters:
+            for name, param in self.named_parameters():
+                if "lora_" in name:
+                    param.requires_grad = False
+
+    def generate(self, *args, **kwargs):
+        # Rely on LoraModel.__getattr__
+        res = self.model.generate(*args, **kwargs)  # type: ignore
+        # TODO(EricLBuehler): Evaluate effectiveness and performance degradation
+        self._freeze_all_adapters()
```

Comment:
We found that the adapters were being reactivated somehow during training after generate, and this was a way to ensure that they are always frozen, as that is very important. Is the apparent reactivation something that should be happening?

Reply:
This doesn't sound good. Instead of trying to fix it here, we should figure out the cause of this reactivation. Do you find it still happens with the latest version of this branch?